### DISCUSSION TITLE
Error in deployment: version ‘GLIBC_2.27’ not found

### DISCUSSION THREAD

**Discussion by RyanGroch** [vercel/community#2105] [Help] [answered]
### Summary

I am currently working with Next.js. I’m doing some 3D rendering in an API endpoint, for which I am using the `three.js` and `gl` npm packages. My app currently works fine in development, but cannot deploy to Vercel.

The most notable error message that occurred in the deployment logs is the following:

```
npm ERR! prebuild-install warn install /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /vercel/path0/node_modules/gl/build/Release/webgl.node)
```

It appears to be the case that the issue is with the `gl` package, which requires GLIBC to run properly. From what I’ve researched, the above error occurs when an OS has a version of GLIBC that is too old for Node to support. 

I’ve looked at some of the suggestions from [this Stack Overflow thread](https://stackoverflow.com/questions/74698021/installing-node-lib64-libm-so-6-version-glibc-2-27-not-found-required-by-n), which seems related and recommends the following potential solutions:
1. Installing Node from source code. I am not sure whether this is something I can do on Vercel, and I’m not sure how this would fix the problem in any case.
2. Installing an older version of Node that the OS can support. After seeing deployment fail with Node 18, I attempted switching my project settings to Node 16 first, and then Node 14 after Node 16 failed with the same error messages. Unfortunately, none of these changes fixed the issue.
3. Upgrading the Linux OS. Again, I’m not sure whether any such thing is possible on Vercel.

Some other things I’ve tried, based on looking at similar issues on Vercel’s Github:
1. Deleting the package-lock.json file. This resulted in slightly different error logs, but the underlying issue appears to be the same.
2. Using yarn instead of npm. Ultimately I view this solution to be the same as deleting package-lock.json, as Vercel attempted to use yarn when it could not detect package-lock.json. The logs when using yarn were the same as after I deleted package-lock.json.

So for now I’m rather stuck. Would anyone happen to know a solution to this? Thank you in advance!

### Additional information

The rest of the error logs from the Vercel deployment (using npm) are shown below:

<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-738dd645-7fff-d2e6-bc10-ecfaaa30cd15"><div dir="ltr" style="margin-left:0pt;" align="left">

npm ERR! code 1 |  
-- | --
10:21:26.993 | npm ERR! path /vercel/path0/node_modules/gl
10:21:26.994 | npm ERR! command failed
10:21:26.994 | npm ERR! command sh -c -- prebuild-install \|\| node-gyp rebuild
10:21:26.994 | npm ERR! prebuild-install warn install /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /vercel/path0/node_modules/gl/build/Release/webgl.node)
10:21:26.995 | npm ERR! gyp info it worked if it ends with ok
10:21:26.995 | npm ERR! gyp info using node-gyp@9.3.1
10:21:26.995 | npm ERR! gyp info using node@16.19.0 \| linux \| x64
10:21:26.995 | npm ERR! gyp info find Python using Python version 3.9.8 found at "/usr/local/bin/python3"
10:21:26.995 | npm ERR! gyp http GET https://nodejs.org/download/release/v16.19.0/node-v16.19.0-headers.tar.gz
10:21:26.995 | npm ERR! gyp http 200 https://nodejs.org/download/release/v16.19.0/node-v16.19.0-headers.tar.gz
10:21:26.995 | npm ERR! gyp http GET https://nodejs.org/download/release/v16.19.0/SHASUMS256.txt
10:21:26.995 | npm ERR! gyp http 200 https://nodejs.org/download/release/v16.19.0/SHASUMS256.txt
10:21:26.996 | npm ERR! gyp info spawn /usr/local/bin/python3
10:21:26.996 | npm ERR! gyp info spawn args [
10:21:26.996 | npm ERR! gyp info spawn args   '/vercel/path0/node_modules/node-gyp/gyp/gyp_main.py',
10:21:26.996 | npm ERR! gyp info spawn args   'binding.gyp',
10:21:26.996 | npm ERR! gyp info spawn args   '-f',
10:21:26.996 | npm ERR! gyp info spawn args   'make',
10:21:26.997 | npm ERR! gyp info spawn args   '-I',
10:21:26.997 | npm ERR! gyp info spawn args   '/vercel/path0/node_modules/gl/build/config.gypi',
10:21:26.997 | npm ERR! gyp info spawn args   '-I',
10:21:26.997 | npm ERR! gyp info spawn args   '/vercel/path0/node_modules/node-gyp/addon.gypi',
10:21:26.997 | npm ERR! gyp info spawn args   '-I',
10:21:26.997 | npm ERR! gyp info spawn args   '/vercel/.cache/node-gyp/16.19.0/include/node/common.gypi',
10:21:26.997 | npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
10:21:26.997 | npm ERR! gyp info spawn args   '-Dvisibility=default',
10:21:26.997 | npm ERR! gyp info spawn args   '-Dnode_root_dir=/vercel/.cache/node-gyp/16.19.0',
10:21:26.997 | npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/vercel/path0/node_modules/node-gyp',
10:21:26.998 | npm ERR! gyp info spawn args   '-Dnode_lib_file=/vercel/.cache/node-gyp/16.19.0/<(target_arch)/node.lib',
10:21:26.998 | npm ERR! gyp info spawn args   '-Dmodule_root_dir=/vercel/path0/node_modules/gl',
10:21:26.998 | npm ERR! gyp info spawn args   '-Dnode_engine=v8',
10:21:26.998 | npm ERR! gyp info spawn args   '--depth=.',
10:21:26.998 | npm ERR! gyp info spawn args   '--no-parallel',
10:21:26.998 | npm ERR! gyp info spawn args   '--generator-output',
10:21:26.998 | npm ERR! gyp info spawn args   'build',
10:21:26.998 | npm ERR! gyp info spawn args   '-Goutput_dir=.'
10:21:26.998 | npm ERR! gyp info spawn args ]
10:21:26.998 | npm ERR! Package xi was not found in the pkg-config search path.
10:21:26.999 | npm ERR! Perhaps you should add the directory containing `xi.pc'
10:21:26.999 | npm ERR! to the PKG_CONFIG_PATH environment variable
10:21:26.999 | npm ERR! No package 'xi' found
10:21:26.999 | npm ERR! gyp: Call to 'pkg-config --libs-only-L --libs-only-other x11 xi xext' returned exit status 1 while in angle/src/angle.gyp. while loading dependencies of binding.gyp while trying to load binding.gyp
10:21:26.999 | npm ERR! gyp ERR! configure error
10:21:26.999 | npm ERR! gyp ERR! stack Error: `gyp` failed with exit code: 1
10:21:26.999 | npm ERR! gyp ERR! stack     at ChildProcess.onCpExit (/vercel/path0/node_modules/node-gyp/lib/configure.js:325:16)
10:21:26.999 | npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
10:21:27.001 | npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)
10:21:27.002 | npm ERR! gyp ERR! System Linux 5.10.173-154.642.amzn2.x86_64
10:21:27.002 | npm ERR! gyp ERR! command "/node16/bin/node" "/vercel/path0/node_modules/.bin/node-gyp" "rebuild"
10:21:27.002 | npm ERR! gyp ERR! cwd /vercel/path0/node_modules/gl
10:21:27.002 | npm ERR! gyp ERR! node -v v16.19.0
10:21:27.003 | npm ERR! gyp ERR! node-gyp -v v9.3.1
10:21:27.003 | npm ERR! gyp ERR! not ok
10:21:27.003 |  
10:21:27.003 | npm ERR! A complete log of this run can be found in:
10:21:27.004 | npm ERR!     /vercel/.npm/_logs/2023-04-17T15_21_04_359Z-debug-0.log
10:21:27.037 | Error: Command "npm install" exited with 1
10:21:27.752 | Deployment completed
10:21:27.319 | BUILD_UTILS_SPAWN_1: Command "npm install" exited with 1

</div><br /></b>
</body>
</html>

**Comment by amyegan** [comment, marked as answer]
Thanks for reaching out about this, @Thrasymachuss. We're currently using GLIBC version 2.26. This is a limitation of our upstream provider, so we can't update it at the moment. 😞 

As a workaround, you could downgrade to a package that allows for version 2.26. Alternatively, you can use [Vercel CLI](https://vercel.com/docs/cli) to build locally and then upload the prebuilt output. For a more automatic deployment process, you can combine the CLI with Github Actions: https://vercel.com/guides/how-can-i-use-github-actions-with-vercel

**Reply by i-am-the-slime** [reply]
@amyegan Do you know of any plans to use a recent version of Amazon Linux on Vercel's side?
